pkg/reconciler/broker/broker.go

```diff
@@ -0,0 +1,499 @@
+	// TODO is this the best way to reconcile triggers when their broker is
+	// being deleted?
+	if err := r.reconcileTriggers(ctx, b); err != nil {
+		logger.Error("Problem reconciling triggers", zap.Error(err))
+		return fmt.Errorf("failed to reconcile triggers: %w", err)
+	}
+
+	if err := r.deleteDecouplingTopicAndSubscription(ctx, b); err != nil {
+		return fmt.Errorf("Failed to delete Pub/Sub topic: %w", err)
+	}
+
+	// TODO should we also delete the broker from the config? Maybe better
+	// to keep the targets in place if there are still triggers. But we can
+	// update the status to UNKNOWN and remove address etc. Maybe need a new
+	// status DELETED or a deleted timestamp that can be used to clean up later.
+
+	return pkgreconciler.NewEvent(corev1.EventTypeNormal, brokerFinalized, "Broker finalized: \"%s/%s\"", b.Namespace, b.Name)
+
+}
+
+func (r *Reconciler) reconcileBroker(ctx context.Context, b *brokerv1beta1.Broker) error {
+	logger := logging.FromContext(ctx)
+	logger.Debug("Reconciling Broker", zap.Any("broker", b))
+	b.Status.InitializeConditions()
+	b.Status.ObservedGeneration = b.Generation
+
+	// Create decoupling topic and pullsub for this broker. Ingress will push
+	// to this topic and fanout will pull from the pull sub.
+	if err := r.reconcileDecouplingTopicAndSubscription(ctx, b); err != nil {
+		return fmt.Errorf("Decoupling topic reconcile failed: %w", err)
+	}
+
+	//TODO in a webhook annotate the broker object with its ingress details
+	// Route to shared ingress with namespace/name in the path as the broker
+	// identifier.
+	b.Status.SetAddress(&apis.URL{
+		Scheme: "http",
+		Host:   names.ServiceHostName(ingressServiceName, system.Namespace()),
+		Path:   fmt.Sprintf("/%s/%s", b.Namespace, b.Name),
+	})
+
+	// Verify the ingress service is healthy via endpoints.
+	ingressEndpoints, err := r.endpointsLister.Endpoints(system.Namespace()).Get(ingressServiceName)
+	if err != nil {
+		logger.Error("Problem getting endpoints for ingress", zap.String("namespace", system.Namespace()), zap.Error(err))
+		b.Status.MarkIngressFailed("ServiceFailure", "%v", err)
+		return err
+	}
+	b.Status.PropagateIngressAvailability(ingressEndpoints)
+
+	r.targetsConfig.MutateBroker(b.Namespace, b.Name, func(m config.BrokerMutation) {
+		m.SetID(string(b.UID))
+		m.SetAddress(b.Status.Address.URL.String())
+		m.SetDecoupleQueue(&config.Queue{
+			Topic:        resources.GenerateDecouplingTopicName(b),
+			Subscription: resources.GenerateDecouplingSubscriptionName(b),
+		})
+		if b.Status.IsReady() {
+			m.SetState(config.State_READY)
+		} else {
+			m.SetState(config.State_UNKNOWN)
+		}
+	})
+
+	// Update config map
+	// TODO should this happen in a defer so there's an update regardless of
+	// error status, or only when reconcile is successful?
+	r.flagTargetsForUpdate()
+	return nil
+}
+
+func (r *Reconciler) reconcileDecouplingTopicAndSubscription(ctx context.Context, b *brokerv1beta1.Broker) error {
```

Comment:
nit: this seems very similar to the reconcileRetryTopicAndSubscriptions. Can this be factored out to be a  common helper function?

Reply:
It seems similar, and I agree this would be worthwhile, especially when we add updates, but there are enough differences and dependencies that it would be some work to factor out. I created #865 to track updating existing topics and subs with the correct configs and added a note to refactor these methods then.